File: typespec/admin/auth/totp.ts
import type {
  TOTPCode,
  TOTPConfiguration,
  TOTPEnrollmentToken,
  TOTPManualEntryKey,
  TOTPRecoveryCode,
  TOTPRecoveryCodeCount,
} from "../../common/authentication.ts";
import {
  isOpaqueToken,
  isTOTPCode,
  isTOTPRecoveryCode,
} from "../../common/authentication.ts";
import type {
  AdminLoginChallengeToken,
  AuthenticatedSessionResponse,
} from "../common/types.ts";

export interface StartTOTPEnrollmentResponse {
  totp_enrollment_token: TOTPEnrollmentToken;
  provisioning_uri: string;
  manual_entry_key: TOTPManualEntryKey;
  configuration: TOTPConfiguration;
  expires_at: string;
}

export interface ConfirmTOTPEnrollmentRequest {
  totp_enrollment_token: TOTPEnrollmentToken;
  totp_code: TOTPCode;
}

export function validateConfirmTOTPEnrollmentRequest(
  request: ConfirmTOTPEnrollmentRequest,
): string[] {
  const fields: string[] = [];
  if (!isOpaqueToken(request.totp_enrollment_token)) {
    fields.push("totp_enrollment_token");
  }
  if (!isTOTPCode(request.totp_code)) {
    fields.push("totp_code");
  }
  return fields;
}

export interface ConfirmTOTPEnrollmentResponse {
  recovery_codes: TOTPRecoveryCode[];
}

export interface VerifyRecoveryCodeRequest {
  login_challenge_token: AdminLoginChallengeToken;
  recovery_code: TOTPRecoveryCode;
}

export function validateVerifyRecoveryCodeRequest(
  request: VerifyRecoveryCodeRequest,
): string[] {
  const fields: string[] = [];
  if (!isOpaqueToken(request.login_challenge_token)) {
    fields.push("login_challenge_token");
  }
  if (!isTOTPRecoveryCode(request.recovery_code)) {
    fields.push("recovery_code");
  }
  return fields;
}

export interface VerifyRecoveryCodeResponse
  extends AuthenticatedSessionResponse {
  remaining_recovery_codes: TOTPRecoveryCodeCount;
}

export interface RegenerateTOTPRecoveryCodesResponse {
  recovery_codes: TOTPRecoveryCode[];
}

File: typespec/hub/auth/totp.go
package auth

import (
	"time"

	"github.com/vetchium/src/typespec/common"
)

type StartTOTPEnrollmentResponse struct {
	TOTPEnrollmentToken common.TOTPEnrollmentToken `json:"totp_enrollment_token"`
	ProvisioningURI     string                     `json:"provisioning_uri"`
	ManualEntryKey      common.TOTPManualEntryKey  `json:"manual_entry_key"`
	Configuration       common.TOTPConfiguration   `json:"configuration"`
	ExpiresAt           time.Time                  `json:"expires_at"`
}

type ConfirmTOTPEnrollmentRequest struct {
	TOTPEnrollmentToken common.TOTPEnrollmentToken `json:"totp_enrollment_token"`
	TOTPCode            common.TOTPCode            `json:"totp_code"`
}

func (r *ConfirmTOTPEnrollmentRequest) Normalize() {}

func (r ConfirmTOTPEnrollmentRequest) Validate() []string {
	fields := make([]string, 0, 2)
	if !common.IsOpaqueToken(string(r.TOTPEnrollmentToken)) {
		fields = append(fields, "totp_enrollment_token")
	}
	if !common.IsTOTPCode(r.TOTPCode) {
		fields = append(fields, "totp_code")
	}
	return fields
}

type ConfirmTOTPEnrollmentResponse struct {
	RecoveryCodes []common.TOTPRecoveryCode `json:"recovery_codes"`
}

type VerifyRecoveryCodeRequest struct {
	LoginChallengeToken HubLoginChallengeToken  `json:"login_challenge_token"`
	RecoveryCode        common.TOTPRecoveryCode `json:"recovery_code"`
}

func (r *VerifyRecoveryCodeRequest) Normalize() {}

func (r VerifyRecoveryCodeRequest) Validate() []string {
	fields := make([]string, 0, 2)
	if !common.IsOpaqueToken(string(r.LoginChallengeToken)) {
		fields = append(fields, "login_challenge_token")
	}
	if !common.IsTOTPRecoveryCode(r.RecoveryCode) {
		fields = append(fields, "recovery_code")
	}
	return fields
}

type VerifyRecoveryCodeResponse struct {
	AuthenticatedSessionResponse
	RemainingRecoveryCodes common.TOTPRecoveryCodeCount `json:"remaining_recovery_codes"`
}

type RegenerateTOTPRecoveryCodesResponse struct {
	RecoveryCodes []common.TOTPRecoveryCode `json:"recovery_codes"`
}
